File: 2018/day09.py

test = (
  (9, 25),
  (10, 1618),
  (13, 7999),
  (17, 1104),
  (21, 6111),
  (30, 5807)
)

live = (
    (448, 71628),
    (448, 7162800)
)

class Marble(object):
    def __init__(self, value):
        self.value = value
        self.next = self
        self.prev = self

    def add( self, new ):
        if new.value % 23 == 0:
            away = self
            for _ in range(7):
                away = away.prev
# Remove
            away.next.prev = away.prev
            away.prev.next = away.next
            return away.next, new.value + away.value
        else:
            next1 = self.next
            next2 = next1.next
            new.next = next2
            new.prev = next1
            next1.next = new
            next2.prev = new
            return new, 0

def process( npl, nmar ):
    nextval = 0
    latest = Marble(nextval)
    score = [0] * npl
    player = 0

    while latest.value < nmar:
        nextval += 1
        latest, points = latest.add( Marble(nextval) )
        score[player] += points
        player = (player + 1) % npl

    while latest.prev:
        n = latest.next
        latest.prev = None
        latest.next = None
        latest = n
    
    return max(score)

for t in test:
    print t, process(*t)
for t in live:
    print t, process(*t)

File: 2018/day09.cpp
#include <list>

test = (
  (9, 25),
  (10, 1618),
  (13, 7999),
  (17, 1104),
  (21, 6111),
  (30, 5807)
)

test = (
  (10, 1618),
  (10, 16180)
)

live = ((448, 71628),)

data = test

void process( int npl, nmar )
{
    std::list<ring>
}

def process( npl, nmar ):
    ring = [0]
    keep = [0] * npl
    player = 0
    for m in range(1,nmar+1):
        if len(ring) == 1:
            ring = [m] + ring
        elif m % 23 == 0:
            print m, player, m, ring[-7]
            keep[player] += m + ring[-7]
            ring = ring[-6:] + ring[:-7]
        else:
            ring = [m] + ring[2:] + ring[:2]
#        print player+1, m, ring
        player = (player + 1) % npl
    print keep
    print max(keep)

process(*test[0])
process(*test[1])

#for run in live:
#    process(*run)
#    break


# So, you score once every 23.
# 23-1 % 9 = 5  gets 23 and  9
# 46-1 % 9 = 0  gets 46 and 17
# 69-1 % 9 = 6  gets 69 and 11
